Provide `arm64` build of `tsc-dyn.dylib` for Apple Silicon (M1) Macs

Currently, `tsc-dyn.dylib` has only `x86_64` binary.

```
✗ ~ file tsc-dyn.dylib 
tsc-dyn.dylib: Mach-O 64-bit dynamically linked shared library x86_64
```

`tsc-dyn.dylib` can be cross-compiled using the latest Rust toolchain.

I'm seeing something similar. I built `tsc-dyn.dylib` but I see the following when running `make ensure/python`:
```
$ EMACS=/Applications/Emacs.app/Contents/MacOS/Emacs make ensure/python
[...]
Debugger entered--Lisp error: (tsc-lang-abi-too-new 13 (9 . 12) "/Users/ddavis/software/repos/emacs-tree-sitter/lan...")
  tsc--load-language("/Users/ddavis/software/repos/emacs-tree-sitter/lan..." "tree_sitter_python" python)
  (let ((language (tsc--load-language full-path native-symbol-name lang-symbol))) (let* ((key lang-symbol)) (condition-case nil (with-no-warnings (map-put! tree-sitter-languages key language nil)) (map-not-inplace (setq tree-sitter-languages (map-insert tree-sitter-languages key language))))) language)
  (let* ((lang-name (symbol-name lang-symbol)) (fallback-name (replace-regexp-in-string "-" "_" lang-name)) (native-symbol-name (or native-symbol-name (format "tree_sitter_%s" fallback-name))) (files (if file (list file) (cons lang-name (if (string= lang-name fallback-name) nil (list fallback-name))))) (full-path (seq-some #'(lambda (base-name) (locate-file base-name tree-sitter-load-path tree-sitter-load-suffixes)) files))) (if full-path nil (error "Cannot find shared library for language: %S" lang-symbol)) (let ((language (tsc--load-language full-path native-symbol-name lang-symbol))) (let* ((key lang-symbol)) (condition-case nil (with-no-warnings (map-put! tree-sitter-languages key language nil)) (map-not-inplace (setq tree-sitter-languages (map-insert tree-sitter-languages key language))))) language))
  tree-sitter-load(python nil nil)
  (or (alist-get lang-symbol tree-sitter-languages) (tree-sitter-load lang-symbol file native-symbol-name))
  tree-sitter-require(python)
  (condition-case nil (tree-sitter-require lang-symbol) (error (display-warning 'tree-sitter-langs (format "Could not load grammar for `%s', trying to compile..." lang-symbol)) (tree-sitter-langs-compile lang-symbol) (tree-sitter-require lang-symbol)))
  (unwind-protect (condition-case nil (tree-sitter-require lang-symbol) (error (display-warning 'tree-sitter-langs (format "Could not load grammar for `%s', trying to compile..." lang-symbol)) (tree-sitter-langs-compile lang-symbol) (tree-sitter-require lang-symbol))) (tree-sitter-langs--copy-query lang-symbol))
  tree-sitter-langs-ensure(python)
  (progn (require 'tree-sitter-langs) (tree-sitter-langs-ensure 'python))
  eval((progn (require 'tree-sitter-langs) (tree-sitter-langs-ensure 'python)) t)
  command-line-1(("--directory" "/Users/ddavis/software/repos/emacs-tree-sitter/cor..." "--directory" "/Users/ddavis/software/repos/emacs-tree-sitter/lis..." "--directory" "/Users/ddavis/software/repos/emacs-tree-sitter/lan..." "--eval" "(progn (require 'tree-sitter-langs) (tree-sitter-l..."))
  command-line()
  normal-top-level()

make: *** [ensure/python] Error 255
```

I don't really know the ins and outs of tree-sitter intimately, or elisp as I'm still new to both of these but I played around with this all for an hour or two and got as far as manually compiling a parser for arm64 (using gcc not clang) but emacs segfaults when that's done. I've looked at the source for `emacs-tree-sitter` as well as `tree-sitter-langs` and `tree-sitter-c-sharp` and as far as I understand it the `parser.c` is generated from [here](https://github.com/ubolonton/tree-sitter-langs/blob/master/tree-sitter-langs-build.el#L283). Where this is compiled into a binary I don't know. I looked through all 3 repos for "clang" and "ld" (which are the things being used as I mucked around a bit and ended up getting this out)

```
Parser compilation failed.
Stdout: 
Stderr: Undefined symbols for architecture x86_64:
  "_tree_sitter_c_sharp_external_scanner_create", referenced from:
      _tree_sitter_c_sharp.language in parser-fc6bff.o
  "_tree_sitter_c_sharp_external_scanner_deserialize", referenced from:
      _tree_sitter_c_sharp.language in parser-fc6bff.o
  "_tree_sitter_c_sharp_external_scanner_destroy", referenced from:
      _tree_sitter_c_sharp.language in parser-fc6bff.o
  "_tree_sitter_c_sharp_external_scanner_scan", referenced from:
      _tree_sitter_c_sharp.language in parser-fc6bff.o
  "_tree_sitter_c_sharp_external_scanner_serialize", referenced from:
      _tree_sitter_c_sharp.language in parser-fc6bff.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

Error calling (tree-sitter test), exit code is 1
make: *** [Makefile:13: ensure/c-sharp] Error 255
```

So I've hit a dead end.

>Emacs collapses the error message. You can click on the dots ... to expand it. You can also modify/advise tree-sitter-load to catch that signal and print the full error data.

The error was just that `c-sharp.dylib` was wrong architecture. I tried to manually run `tree-sitter generate`

Now I'm also stuck on what looks like the same issue as above.

running `clang -o parser.so -shared src/parser.c -I./src -v`

```
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: arm64-apple-darwin20.3.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
 "/Library/Developer/CommandLineTools/usr/bin/clang" -cc1 -triple arm64-apple-macosx11.0.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -Werror=implicit-function-declaration -emit-obj -mrelax-all -disable-free -disable-llvm-verifier -discard-value-names -main-file-name parser.c -mrelocation-model pic -pic-level 2 -mthread-model posix -mframe-pointer=non-leaf -fno-strict-return -masm-verbose -munwind-tables -target-sdk-version=11.1 -target-cpu vortex -target-feature +v8.3a -target-feature +fp-armv8 -target-feature +neon -target-feature +crc -target-feature +crypto -target-feature +fullfp16 -target-feature +ras -target-feature +lse -target-feature +rdm -target-feature +rcpc -target-feature +zcm -target-feature +zcz -target-feature +sha2 -target-feature +aes -target-abi darwinpcs -fallow-half-arguments-and-returns -dwarf-column-info -debugger-tuning=lldb -target-linker-version 609.8 -v -resource-dir /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I ./src -I/usr/local/include -internal-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include -internal-isystem /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include -internal-externc-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include -internal-externc-isystem /Library/Developer/CommandLineTools/usr/include -Wno-reorder-init-list -Wno-implicit-int-float-conversion -Wno-c99-designator -Wno-final-dtor-non-final-class -Wno-extra-semi-stmt -Wno-misleading-indentation -Wno-quoted-include-in-framework-header -Wno-implicit-fallthrough -Wno-enum-enum-conversion -Wno-enum-float-conversion -fdebug-compilation-dir /Users/theo/Git/emacs-tree-sitter/langs/repos/c-sharp -ferror-limit 19 -fmessage-length 151 -stack-protector 1 -fstack-check -mdarwin-stkchk-strong-link -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fgnuc-version=4.2.1 -fobjc-runtime=macosx-11.0.0 -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -o /var/folders/0l/7j7ds92x343fn8xnlrg1lv180000gp/T/parser-8e0edc.o -x c src/parser.c
clang -cc1 version 12.0.0 (clang-1200.0.32.29) default target arm64-apple-darwin20.3.0
ignoring nonexistent directory "/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 ./src
 /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include
 /Library/Developer/CommandLineTools/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.
 "/Library/Developer/CommandLineTools/usr/bin/ld" -demangle -lto_library /Library/Developer/CommandLineTools/usr/lib/libLTO.dylib -no_deduplicate -dynamic -dylib -arch arm64 -platform_version macos 11.0.0 11.1 -syslibroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -o parser.so -L/usr/local/lib /var/folders/0l/7j7ds92x343fn8xnlrg1lv180000gp/T/parser-8e0edc.o -lSystem /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a
Undefined symbols for architecture arm64:
  "_tree_sitter_c_sharp_external_scanner_create", referenced from:
      _tree_sitter_c_sharp.language in parser-8e0edc.o
  "_tree_sitter_c_sharp_external_scanner_deserialize", referenced from:
      _tree_sitter_c_sharp.language in parser-8e0edc.o
  "_tree_sitter_c_sharp_external_scanner_destroy", referenced from:
      _tree_sitter_c_sharp.language in parser-8e0edc.o
  "_tree_sitter_c_sharp_external_scanner_scan", referenced from:
      _tree_sitter_c_sharp.language in parser-8e0edc.o
  "_tree_sitter_c_sharp_external_scanner_serialize", referenced from:
      _tree_sitter_c_sharp.language in parser-8e0edc.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ## [Unreleased]
 - Added customization option `tsc-dyn-get-from`, which is a list of sources to get the dynamic module `tsc-dyn` from. Its default value is `(:github :compilation)`.
 - Made `tree-sitter-hl`'s region-fontification function fall back to the underlying non-tree-sitter function when called outside of `tree-sitter-hl-mode`. This fixes an issue where `jupyter-repl-mode`'s [input cells are not highlighted](https://github.com/nnicandro/emacs-jupyter/issues/363).
+- Updated `tsc-dyn-get` to download platform-specific binaries (mainly for Apple Silicon).
 
 ## [0.16.1] - 2021-12-11
 - Modified CI pipelines to publish additional pre-built dynamic modules. Their filenames include the platform they are built for. The files without platform in name will eventually be deprecated.

--- a/core/tsc-dyn-get.el
+++ b/core/tsc-dyn-get.el
@@ -77,8 +77,19 @@ this to nil."
     (_ "so")))
 
 (defun tsc-dyn-get--file ()
-  "Return the dynamic module filename, which is system-dependent."
+  "Return the dynamic module filename, which is OS-dependent."
   (format "tsc-dyn.%s" (tsc-dyn-get--ext)))
+
+;;; TODO: Make this correct.
+(defun tsc-dyn-get--system-specific-file ()
+  "Return the dynamic module filename, which is system-dependent."
+  (pcase system-type
+    ('windows-nt "tsc-dyn.x86_64-pc-windows-msvc.dll")
+    ('darwin (if (string-prefix-p "x86_64" system-configuration)
+                 "tsc-dyn.x86_64-apple-darwin.dylib"
+               "tsc-dyn.aarch64-apple-darwin.dylib"))
+    ((or 'gnu 'gnu/linux 'gnu/kfreebsd)
+     "tsc-dyn.x86_64-unknown-linux-gnu.so")))
 
 (defun tsc-dyn-get--log (format-string &rest args)
   (apply #'message (concat "tsc-dyn-get: " format-string) args))
@@ -123,19 +134,24 @@ This function records the downloaded version in the manifest
   (let* ((bin-dir (tsc-dyn-get--dir))
          (default-directory bin-dir)
          (_ (unless (file-directory-p bin-dir) (make-directory bin-dir)))
-         (dyn-file (tsc-dyn-get--file))
-         (gz-file (format "%s.gz" dyn-file))
          (uncompressed? (version< "0.7.0" version))
+         (system-specific? (version<= "0.16.1" version))
+         (local-name (tsc-dyn-get--file))
+         (remote-name (format "%s%s"
+                              (if system-specific?
+                                  (tsc-dyn-get--system-specific-file)
+                                local-name)
+                              (if uncompressed? "" ".gz")))
          (url (format "https://github.com/emacs-tree-sitter/elisp-tree-sitter/releases/download/%s/%s"
-                      version (if uncompressed? dyn-file gz-file))))
+                      version remote-name)))
     (tsc-dyn-get--log "Downloading %s" url)
     (if uncompressed?
-        (tsc-dyn-get--url-copy-file url dyn-file :ok-if-already-exists)
-      (tsc-dyn-get--url-copy-file url gz-file)
-      (when (file-exists-p dyn-file)
-        (delete-file dyn-file))
+        (tsc-dyn-get--url-copy-file url local-name :ok-if-already-exists)
+      (tsc-dyn-get--url-copy-file url remote-name)
+      (when (file-exists-p local-name)
+        (delete-file local-name))
       ;; XXX: Uncompressing with `dired-compress-file' doesn't work on Windows.
-      (dired-compress-file gz-file))
+      (dired-compress-file remote-name))
     (with-temp-file tsc-dyn-get--version-file
       (let ((coding-system-for-write 'utf-8))
         (insert version)))))

